File: packages/sillytavern-compat/src/detect.ts
import type { JsonValue } from "@zhuangsheng/api-client";
import { object } from "./json";
import type { SillyTavernPresetKind } from "./types";

export function detectPresetKind(document: JsonValue): SillyTavernPresetKind {
  if (isRegexArray(document)) return "regex_scripts";
  const value = object(document);
  if (!value) return "unknown";
  if (Array.isArray(value.prompts) && Array.isArray(value.prompt_order)) return "open_ai";
  if (embeddedRegex(value)) return "regex_scripts";
  if (["context", "instruct", "sysprompt", "reasoning", "preset"].some((key) => key in value)) return "master";
  if ("story_string" in value) return "context";
  if ("input_sequence" in value && "output_sequence" in value) return "instruct";
  if ("content" in value && "name" in value) return "system_prompt";
  if (["prefix", "suffix", "separator"].every((key) => key in value)) return "reasoning";
  if (["temp", "top_k", "top_p", "rep_pen"].every((key) => key in value)) return "text_completion";
  return "unknown";
}

export function isRegexArray(value: JsonValue | undefined): boolean {
  return Array.isArray(value) && value.every((entry) => {
    const rule = object(entry);
    return !!rule && typeof rule.findRegex === "string" && typeof rule.replaceString === "string";
  });
}

function embeddedRegex(value: Record<string, JsonValue>) {
  const data = object(value.data);
  const dataExtensions = data && object(data.extensions);
  const extensions = object(value.extensions);
  return isRegexArray(dataExtensions?.regex_scripts) || isRegexArray(extensions?.regex_scripts);
}

File: crates/core/src/compatibility/sillytavern/detect.rs
use serde_json::Value;

use super::SillyTavernPresetKind;

pub fn detect_preset_kind(document: &Value) -> SillyTavernPresetKind {
    if is_regex_array(document) {
        return SillyTavernPresetKind::RegexScripts;
    }
    let Some(object) = document.as_object() else {
        return SillyTavernPresetKind::Unknown;
    };
    if object.get("prompts").is_some_and(Value::is_array)
        && object.get("prompt_order").is_some_and(Value::is_array)
    {
        return SillyTavernPresetKind::OpenAi;
    }
    if ["context", "instruct", "sysprompt", "reasoning", "preset"]
        .iter()
        .any(|key| object.contains_key(*key))
    {
        return SillyTavernPresetKind::Master;
    }
    if object.contains_key("story_string") {
        return SillyTavernPresetKind::Context;
    }
    if object.contains_key("input_sequence") && object.contains_key("output_sequence") {
        return SillyTavernPresetKind::Instruct;
    }
    if object.contains_key("content") && object.contains_key("name") {
        return SillyTavernPresetKind::SystemPrompt;
    }
    if ["prefix", "suffix", "separator"]
        .iter()
        .all(|key| object.contains_key(*key))
    {
        return SillyTavernPresetKind::Reasoning;
    }
    if ["temp", "top_k", "top_p", "rep_pen"]
        .iter()
        .all(|key| object.contains_key(*key))
    {
        return SillyTavernPresetKind::TextCompletion;
    }
    SillyTavernPresetKind::Unknown
}

fn is_regex_array(value: &Value) -> bool {
    value.as_array().is_some_and(|values| {
        values.iter().all(|value| {
            value.get("findRegex").is_some_and(Value::is_string)
                && value.get("replaceString").is_some_and(Value::is_string)
        })
    })
}
